src/stream/src/executor/hash_agg.rs

```diff
@@ -410,68 +428,48 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
 
-                // Create array builders.
-                // As the datatype is retrieved from schema, it contains both group key and
-                // aggregation state outputs.
-                let mut builders = this.info.schema.create_array_builders(this.chunk_size * 2);
-                let mut new_ops = Vec::with_capacity(this.chunk_size * 2);
-
-                // Calculate current outputs, concurrently.
-                let futs = keys_in_batch.into_iter().map(|key| {
+            let futs_of_all_groups = vars
+                .group_change_set
+                .drain()
+                .map(|key| {
                     // Get agg group of the key.
-                    let agg_group = {
-                        let mut ptr: NonNull<_> = vars
-                            .agg_group_cache
-                            .get_mut(&key)
-                            .expect("changed group must have corresponding AggGroup")
-                            .into();
-                        // SAFETY: `key`s in `keys_in_batch` are unique by nature, because they're
-                        // from `group_change_set` which is a set.
-                        unsafe { ptr.as_mut() }
-                    };
-                    async {
-                        let curr_outputs = agg_group.get_outputs(&this.storages).await?;
-                        Ok::<_, StreamExecutorError>((key, agg_group, curr_outputs))
-                    }
+                    let mut ptr: NonNull<_> = vars
+                        .agg_group_cache
+                        .get_mut(&key)
+                        .expect("changed group must have corresponding AggGroup")
+                        .into();
+                    // SAFETY: `key`s in `keys_in_batch` are unique by nature, because they're
+                    // from `group_change_set` which is a set.
+                    unsafe { ptr.as_mut() }
+                })
+                .map(|agg_group| async {
+                    // Get agg outputs and build change.
+                    let curr_outputs = agg_group.get_outputs(&this.storages).await?;
+                    let change = agg_group.build_change(curr_outputs);
+                    Ok::<_, StreamExecutorError>(change)
                 });
-                let outputs_in_batch: Vec<_> = stream::iter(futs)
-                    .buffer_unordered(10)
-                    .fuse()
-                    .try_collect()
-                    .await?;
-
-                for (_key, agg_group, curr_outputs) in outputs_in_batch {
-                    if let Some(change) = agg_group.build_change(curr_outputs) {
-                        agg_group.apply_change_to_builders(&change, &mut builders, &mut new_ops);
-                        agg_group.apply_change_to_result_table(&change, &mut this.result_table);
+
+            // TODO(rc): figure out a more reasonable concurrency limit.
+            const MAX_CONCURRENT_TASKS: usize = 100;
```

Comment:
Need some discussion on this number. Previously we use `.buffer_unordered(10)`, but IIUC there're actually `this.chunk_size` tasks running simultaneously. Since the results of all futures in `futs` are collected when they all finished, the buffer with size 10 is meaningless.